casper/src/main/resources/RevVault.rho

```diff
@@ -84,27 +85,51 @@ in {
             } |
 
             contract _getOrCreate(@revAddress, constructor, retCh) = {
-              new revAddressValidCh, revAddressValidEitherCh, getOrCreate in {
+              new revAddressValidCh,
+                  revAddressValidEitherCh,
+                  getOrCreate
+              in {
                 RevAddress!("validate", revAddress, *revAddressValidCh) |
                 @Either!("fromNillableError <-", *revAddressValidCh, *revAddressValidEitherCh) |
                 @Either!("flatMap <-", *revAddressValidEitherCh, *getOrCreate, *retCh) |
                 for (_, ret <- getOrCreate) {
-                  for (@vaultMap <- vaultMapStore) {
+                  // At first peek for `vaultMap` to check if vault (address) exists
+                  // - peek ensures that for existing vaults we don't have a block conflict
+                  for (@vaultMap <<- vaultMapStore) {
                     match vaultMap.get(revAddress) {
                       Nil => {
-                        new resCh in {
-                          constructor!(*resCh) |
-                          for (@eitherVault <- resCh) {
-                            match eitherVault {
-                              (true, vault) => vaultMapStore!(vaultMap.set(revAddress, vault))
-                              (false,  _    ) => vaultMapStore!(vaultMap)
-                            } |
-                            ret!(eitherVault)
+                        // Vault is not found, create new and save it in the `vaultMap`
+                        // - this means two blocks with new vaults cannot be merged
+                        for (@vaultMap <- vaultMapStore) {
+                          // We need to check again because `vaultMap` is not locked with peek
+                          match vaultMap.get(revAddress) {
+                            Nil => {
+                              new resCh in {
+                                constructor!(*resCh) |
+                                for (@eitherVault <- resCh) {
+                                  match eitherVault {
+                                    (true, vault) => {
+                                      vaultMapStore!(vaultMap.set(revAddress, vault))
+                                    }
+                                  } |
```

Comment:
You still need the `false` case to put `vaultMap` back into `vaultMapStore`.  Otherwise `vaultMap` goes out of scope and is lost.  Please add a test for this case.